gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+		uint8_t fib_id;
+
+		prefix_len = sizeof(struct in6_addr) * 8;
+
+		gt_fib_present = rte_lpm6_is_rule_present(
+			ltbl->lpm6, gt_addr.ip.v6.s6_addr,
+			prefix_len, &fib_id);
+
+		ret = lpm_lookup_ipv6(ltbl->lpm6, gt_addr.ip.v6.s6_addr);
+		if (ret >= 0)
+			gt_prefix_fib = &ltbl->fib_tbl6[ret];
+	} else
+		rte_panic("Unexpected condiction: unknown IP type %hu at %s!\n",
+			gt_addr.proto, __func__);
+
+	if (gt_prefix_fib == NULL)
+		return NULL;
+
+	/* It's the FIB entry for the Grantor. */
+	if (gt_fib_present)
+		return gt_prefix_fib;
+
+	if ((gt_prefix_fib->action != GK_FWD_GATEWAY_BACK_NET &&
+			gt_prefix_fib->action != GK_FWD_NEIGHBOR_BACK_NET)) {
+		return NULL;
+	}
+
+	gt_fib = add_prefix_fib(&gt_addr, prefix_len, gk_conf);
+	if (gt_fib == NULL)
+		return NULL;
+
+	gt_fib->action = GK_FWD_GRANTOR;
+	flow = &gt_fib->u.grantor.flow;
+
+	if (gt_addr.proto == ETHER_TYPE_IPv4) {
+		flow->proto = ETHER_TYPE_IPv4;
+		flow->f.v4.src = gk_conf->net->back.ip4_addr.s_addr;
+		flow->f.v4.dst = gt_addr.ip.v4.s_addr;
+	} else {
+		flow->proto = ETHER_TYPE_IPv6;
+		rte_memcpy(flow->f.v6.src,
+			gk_conf->net->back.ip6_addr.s6_addr,
+			sizeof(flow->f.v6.src));
+		rte_memcpy(flow->f.v6.dst,
+			gt_addr.ip.v6.s6_addr, sizeof(flow->f.v6.dst));
+	}
+
+	/* Fill up the @next_fib field in the @gt_fib. */
+
+	/*
+	 * The Grantor server is not an neighbor.
+	 * So the @next_fib needs to point to the
+	 * responsbile gateway entry.
+	 */
+	if (gt_prefix_fib->action == GK_FWD_GATEWAY_BACK_NET) {
+		gt_fib->u.grantor.next_fib = gt_prefix_fib;
+		gt_fib->u.grantor.next_fib->ref_cnt++;
+		gt_fib->u.grantor.eth_cache =
+			gt_prefix_fib->u.gateway.eth_cache;
+
+		/* TODO Add concurrency control. */
+		gt_fib->u.grantor.eth_cache->ref_cnt++;
+
+		goto out;
+	}
+
+	/* The Grantor server is an neighbor. */
+	if (gt_addr.proto == ETHER_TYPE_IPv4) {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh,
+			&gt_addr.ip.v4.s_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh.hash_table,
+				&gt_addr.ip.v4.s_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				rte_lpm_delete(ltbl->lpm,
+					ntohl(gt_addr.ip.v4.s_addr),
+					(sizeof(struct in_addr) * 8));
+				return NULL;
+			}
+		}
+
+	} else {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh6,
+			gt_addr.ip.v6.s6_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh6.hash_table,
+				gt_addr.ip.v6.s6_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh6.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				rte_lpm6_delete(ltbl->lpm6,
+					gt_addr.ip.v6.s6_addr,
+					(sizeof(struct in6_addr) * 8));
+				return NULL;
+			}
+		}
```

Comment:
Is there any way that you can reduce this `lookup_ether_cache()` and the resulting nested conditional statements into a function? Since it is essentially used four times: twice in this function and twice in `init_gateway_fib()`.